### PR TITLE
[13.0] stock_dynamic_routing: fix compatibility with stock_move_source_relocate module

### PR DESCRIPTION
**Pull request by jbaudoux**
After applying the dynamic routing, call action_assign on all moves (also those non re-routed) to ensure that other hooks on that method get the complete list of moves.
Once all moves have been re-routed, propagate the changes to the destination moves.

Before this change, if you have product with 10 in stock and create a pick+ship for 15, with a dynamic routing on the pick to pull from location A to location Output/Area1, it will split the pick stock move in 10 and 5, update the destination location on the move of 10, propagate the change of source location on the ship move that gets split also. Then if you have a relocate rule for non-assigned pick from location A to Output/Area1, the pick move of 5 is re-routed, then the source location of the ship move of 5 is updated. When printing the delivery slip, you have 2 lines for the same product as the ship move was split.

After this change, it will split the pick stock move in 10 and 5, update the locations on the move of 10, then the move 5 gets relocated. Finally it propagates the change on the ship move that sees that both orig moves
are going to the same destination location and then does not split the ship move. When printing the delivery slip, you have 1 line for the product as the ship move was not split.

cc @jgrandguillaume @guewen

### CODE REVIEW

**Comment by jbaudoux** [comment]
TODO add test for described case

**Comment by jgrandguillaume** [comment]
Hi @jbaudoux ,

That's a nice one, well done. Look, for what I'm able to judge (and after some headache in the understanding) I think this change is ok.

I would really appreciate you write a test case as you mentioned before we accept to merge it.

**Comment by jbaudoux** [comment]
> TODO add test for described case

done

**Comment by jbaudoux** [comment]
TODO refactor & add test for insert move case

**Comment by jbaudoux** [comment]
> TODO refactor & add test for insert move case

refactored code for insert move to prevent recurring calls and properly propagate re-routing. Test added for that case

**Review by simahawk — COMMENTED** [review]
LG and seems to make sense but as I don't have a deep knowledge on routing I wait for others to approve.

**Review by sebalix — APPROVED** [review]
Seems OK, I suppose :)

**Review by mmequignon — APPROVED** [review]
Code looks good

**Comment by jbaudoux** [comment]
I fixed the push routing to ensure putaway is applied on the initial move

**Comment by sebalix** [comment]
@jbaudoux one test is failing on `odoo/odoo`
```python
FAIL: LocationContentTransferSetDestinationAllCase.test_set_destination_all_with_partially_available_move_with_ancestor
Traceback (most recent call last):
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/tests/test_location_content_transfer_set_destination_all.py", line 161, in test_set_destination_all_with_partially_available_move_with_ancestor
    params={"location_id": self.content_loc.id, "barcode": sub_shelf1.barcode},
  File "/home/travis/build/OCA/wms/setup/shopfloor_base/odoo/addons/shopfloor_base/services/service.py", line 27, in dispatch
    return super().dispatch(method_name, *args, params=params)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/odoo/addons/rest_log/components/service.py", line 36, in dispatch
    return super().dispatch(method_name, *args, params=params)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/odoo/addons/base_rest/components/service.py", line 158, in dispatch
    res = method(*args, **secure_params)
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/odoo/addons/base_rest/restapi.py", line 61, in response_wrap
    response = f(*args, **kw)
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/services/location_content_transfer.py", line 477, in set_destination_all
    self._set_all_destination_lines_and_done(pickings, move_lines, scanned_location)
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/services/location_content_transfer.py", line 434, in _set_all_destination_lines_and_done
    stock.validate_moves(move_lines.move_id)
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/actions/stock.py", line 30, in validate_moves
    moves_todo.extract_and_action_done()
  File "/home/travis/build/OCA/wms/setup/shopfloor/odoo/addons/shopfloor/models/stock_move.py", line 104, in extract_and_action_done
    assert new_picking.state == "assigned"
AssertionError
```
I restart the Travis job to check as all was green on `OCA/OCB`...

File: stock_dynamic_routing/models/stock_move.py
Comment (sebalix) [inline stock_dynamic_routing/models/stock_move.py:246]: The `method` selection field has no default value nor is required, so we could have no value here
```suggestion
            if routing_rule.method != "pull":
```

Comment (sebalix) [reply stock_dynamic_routing/models/stock_move.py:246]: Or having no value means it should be applied on both push & pull?

**Comment by sebalix** [comment]
Test fails on `odoo/odoo` because of this new commit https://github.com/odoo/odoo/commit/591a726f4901b863f7a13581942e7e6b352c1728 still not synced on `OCA/OCB`, checking why.

**Comment by sebalix** [comment]
A fix has been pushed here: https://github.com/OCA/wms/pull/275

**Comment by sebalix** [comment]
@jbaudoux I restarted the build, should be OK now, Odoo made a fix regarding the mentioned issue here: https://github.com/odoo/odoo/pull/75019

**Comment by sebalix** [comment]
@jbaudoux tests on `stock_move_source_relocate_dynamic_routing` are broken

**Comment by jbaudoux** [comment]
@sebalix Thanks for the info. I reverted your last suggestion (== "push" to != "pull") as it was the only change I did since the tests were green

**Review by jgrandguillaume — APPROVED** [review]
LGTM

**Pull request merged**
